### PR TITLE
[WIP] PIM-5558

### PR DESCRIPTION
**Pull request by grena**
Todo list 
- [ ] `InputDecorator` : rename every method which contains 'field'. Input should be used instead.
- [ ] `InputDecorator` : `findCompoundField` should be renamed.
- [x] `WebUserContext` : line 850, if/if/else de la mort qui tue.
- [x] PEF et variant group form : a "select2 result dropdown"  is added to the dom when you change the current tab and you reopen it.
- [x] `AttributeAdder` refactoring (code duplication)
- [ ] Decorator : `getSession` is deprecated but we need it. We should find a workaround.
- [ ] Seems that non scopable attributes are displayed like scopable ones.

TIP : The namespace of the decorator are not consistent.

### CODE REVIEW

File: features/Context/AssertionContext.php [hunk 1]
```diff
@@ -161,27 +161,19 @@ public function iShouldSeeTheFields($fields)
...
         $currencies = array_map('trim', $currencies);
-        $priceField = $priceLabelField->getParent();
 
         foreach ($currencies as $currency) {
-            if (null === $priceField->find('css', sprintf('.controls input[value="%s"]', $currency))) {
-                throw $this->createExpectationException(
-                    sprintf('Expecting to see the currency "%s" on price field "%s".', $currency, $field)
-                );
-            }
+            $this->getCurrentPage()->getElement('Attribute inputs')->findCurrencyInput($inputLabel, $currency);
```
Comment (juliensnz) [inline features/Context/AssertionContext.php]: This feels more natural:
`$this->getCurrentPage()->getAttributeTab()->getAttributeField($inputLabel)->getCurrencies();`
and is possible with the current decorator engine :)

Comment (juliensnz) [inline features/Context/AssertionContext.php]: `getAttributeTab` will return the attribute tab and could be a AttributeCapablePageDecorator or a simpleDecorator
`getAttributeField` or `getField` could return the field container
`getCurrencies` should return the list of currencies

File: features/Context/Page/VariantGroup/Edit.php [hunk 2]
```diff
@@ -48,32 +67,7 @@ public function __construct($session, $pageFactory, $parameters = [])
      */
     public function findField($name)
```
Comment (juliensnz) [inline features/Context/Page/VariantGroup/Edit.php]: the goal of the decorator is to drop all the code from pages. So you can drop this method from this page and call it directly in the context. If not possible for now, you can call keep it here but add a todo :)

File: features/Context/Page/VariantGroup/Edit.php [hunk 3]
```diff
@@ -103,6 +97,14 @@ public function getRemoveLinkFor($field)
     }
 
     /**
+     * {@inheritdoc}
+     */
+    public function save()
+    {
+        $this->getElement('Navbar buttons')->asynchronousSave('Save');
```
Comment (juliensnz) [inline features/Context/Page/VariantGroup/Edit.php]: From here we should not know that it's asynchronous. So the call should be: `$this->getCurrentPage()->save()` with a SaveActionCapablePageDecorator

File: features/Context/WebUser.php [hunk 4]
```diff
@@ -855,7 +855,9 @@ public function iClickOnTheField($field)
     public function iShouldSeeAvailableAttributesInGroup($not, $attributes, $group)
     {
         foreach ($this->listToArray($attributes) as $attribute) {
-            $element = $this->getCurrentPage()->findAvailableAttributeInGroup($attribute, $group);
+            $element = $this->getCurrentPage()
+                ->getElement('Add attributes button')
```
Comment (juliensnz) [inline features/Context/WebUser.php]: `Add attributes`

File: features/Context/WebUser.php [hunk 5]
```diff
@@ -855,7 +855,9 @@ public function iClickOnTheField($field)
     public function iShouldSeeAvailableAttributesInGroup($not, $attributes, $group)
```
Comment (juliensnz) [inline features/Context/WebUser.php]: should be moved in a proper context

File: features/Context/WebUser.php [hunk 6]
```diff
@@ -887,7 +889,9 @@ public function iShouldSeeAvailableAttributesInGroup($not, $attributes, $group)
      */
     public function iAddAvailableAttributes($attributes)
```
Comment (juliensnz) [inline features/Context/WebUser.php]: should be moved in a proper context

File: features/Context/WebUser.php [hunk 7]
```diff
@@ -887,7 +889,9 @@ public function iShouldSeeAvailableAttributesInGroup($not, $attributes, $group)
...
     {
-        $this->getCurrentPage()->addAvailableAttributes($this->listToArray($attributes));
+        $this->getCurrentPage()
+            ->getElement('Add attributes button')
```
Comment (juliensnz) [inline features/Context/WebUser.php]: `Add attribute`

File: features/Context/WebUser.php [hunk 8]
```diff
@@ -887,7 +889,9 @@ public function iShouldSeeAvailableAttributesInGroup($not, $attributes, $group)
...
+            ->addAvailableAttributes($this->listToArray($attributes));
```
Comment (juliensnz) [inline features/Context/WebUser.php]: no mandatory but could be reworked to do a `addAvailableAttribute` multiple times.

File: features/Pim/Behat/Decorator/Attribute/AttributeAdderDecorator.php [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+use Behat\Mink\Element\NodeElement;
+use Context\Spin\SpinCapableTrait;
+use Pim\Behat\Decorator\ElementDecorator;
+
+/**
+ * @author    Adrien Pétremann <adrien.petremann@akeneo.com>
+ * @copyright 2016 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class AttributeAdderDecorator extends ElementDecorator
```
Comment (juliensnz) [inline features/Pim/Behat/Decorator/Attribute/AttributeAdderDecorator.php]: Could be renamed into SelectAttributeDecorator (for mass edit for example)

File: features/Pim/Behat/Decorator/Attribute/AttributeAdderDecorator.php [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+{
+    use SpinCapableTrait;
+
+    protected $selectors = [
+        'Available attributes button'     => ['css' => '.add-attribute a.select2-choice'],
```
Comment (juliensnz) [inline features/Pim/Behat/Decorator/Attribute/AttributeAdderDecorator.php]: it should be `'button' => 'a.select2-choice'` as we already decorate a '.add-attribute` element

Comment (juliensnz) [inline features/Pim/Behat/Decorator/Attribute/AttributeAdderDecorator.php]: you can also drop the css part

File: features/Pim/Behat/Decorator/Attribute/AttributeAdderDecorator.php [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+        $selector->click();
+
+        $list = $this->spin(function () {
+            return $this->find('css', $this->selectors['Available attributes list']['css']);
+        }, 'Cannot find the attribute list element');
+
+        // We NEED to fill the search field with jQuery to avoid the TAB key press (because of mink),
+        // because select2 selects the first element on TAB key press.
+        $this->getSession()->evaluateScript(
+            "jQuery('" . $searchSelector . "').val('" . $attribute . "').trigger('input');"
```
Comment (juliensnz) [inline features/Pim/Behat/Decorator/Attribute/AttributeAdderDecorator.php]: maybe use a sprintf

File: features/Pim/Behat/Decorator/Attribute/AttributeAdderDecorator.php [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+        );
+
+        $groupLabels = $this->spin(function () use ($list, $group) {
+            return $list->findAll('css', sprintf('li .group-label:contains("%s"), li.select2-no-results', $group));
```
Comment (juliensnz) [inline features/Pim/Behat/Decorator/Attribute/AttributeAdderDecorator.php]: like here :)

File: features/Pim/Behat/Decorator/Attribute/AttributeAdderDecorator.php [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+        if ('No matches found' !== $text) {
+            foreach ($groupLabels as $groupLabel) {
+                $li = $groupLabel->getParent();
+                $results[$li->find('css', '.attribute-label')->getText()] = $li;
+            }
+        }
+
+        // Close select2
+        $this->getSession()->evaluateScript(
+            "jQuery('" . $this->selectors['Select2 dropmask']['css'] . "').click();"
```
Comment (juliensnz) [inline features/Pim/Behat/Decorator/Attribute/AttributeAdderDecorator.php]: same here

File: features/Pim/Behat/Decorator/Attribute/AttributeAdderDecorator.php [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+        return isset($results[$attribute]) ? $results[$attribute] : null;
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * TODO: Used with the new 'add-attributes' module. The method should be in the Form parent
+     * when legacy stuff is removed.
+     */
+    public function addAvailableAttributes(array $attributes = [])
```
Comment (juliensnz) [inline features/Pim/Behat/Decorator/Attribute/AttributeAdderDecorator.php]: would be more srp if it took only one attribute at a time

File: features/Pim/Behat/Decorator/InputDecorator.php [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+ * @author    Adrien Pétremann <adrien.petremann@akeneo.com>
+ * @copyright 2016 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class InputDecorator extends ElementDecorator
+{
+    use SpinCapableTrait;
+
+    protected $selectors = [
+        'Input label' => ['css' => '.field-container header label:contains("%s")'],
```
Comment (juliensnz) [inline features/Pim/Behat/Decorator/InputDecorator.php]: you can drop the css key

File: features/Pim/Behat/Decorator/InputDecorator.php [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+
+    /**
+     * @param string $label
+     * @param bool   $copy
+     *
+     * @throws ElementNotFoundException
+     *
+     * @return NodeElement
+     */
+    public function findField($label, $copy = false)
```
Comment (juliensnz) [inline features/Pim/Behat/Decorator/InputDecorator.php]: the copy part should be dropped as we acknowledged that flag operator should not be used

File: features/Pim/Behat/Decorator/InputDecorator.php [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+        $subContainer = $this->spin(function () use ($label, $copy) {
+            return $this->findFieldContainer($label)
+                ->find('css', $copy ? '.copy-container .form-field' : '.form-field');
+        });
+
+        $field = $this->spin(function () use ($subContainer) {
+            return $subContainer->find('css', $this->selectors['Text inputs']['css']);
+        });
+
+        return $field;
```
Comment (juliensnz) [inline features/Pim/Behat/Decorator/InputDecorator.php]: this method should return the field container

File: features/Pim/Behat/Decorator/InputDecorator.php [hunk 18]
```diff
@@ -0,0 +1,119 @@
...
+    /**
+     * Find field container
+     *
+     * @param string $label
+     *
+     * @throws ElementNotFoundException
+     *
+     * @return NodeElement
+     */
+    public function findFieldContainer($label)
```
Comment (juliensnz) [inline features/Pim/Behat/Decorator/InputDecorator.php]: yeah, a lot of code duplicated here. As we rework this stuff we should clean it and not do the old errors again

File: features/Pim/Behat/Decorator/InputDecorator.php [hunk 19]
```diff
@@ -0,0 +1,119 @@
...
+     * Find a compound field
+     *
+     * @param NodeElement $fieldContainer
+     * @param             $currency
+     *
+     * @throws ElementNotFoundException
+     *
+     * @return NodeElement
+     */
+    protected function findCompoundField($fieldContainer, $currency)
```
Comment (juliensnz) [inline features/Pim/Behat/Decorator/InputDecorator.php]: this kind of method is not relevant anymore IMHO

File: features/Pim/Behat/Decorator/InputDecorator.php [hunk 20]
```diff
@@ -0,0 +1,119 @@
...
+        return $input;
+    }
+
+    /**
+     * @param string $inputLabel
+     * @param string $currency
+     *
+     * @return NodeElement
+     */
+    public function findCurrencyInput($inputLabel, $currency)
```
Comment (juliensnz) [inline features/Pim/Behat/Decorator/InputDecorator.php]: not in this decorator :)

File: features/Pim/Behat/Decorator/Navbar/ButtonDecorator.php [hunk 21]
```diff
@@ -0,0 +1,67 @@
...
+use Context\Spin\SpinCapableTrait;
+use Context\Spin\TimeoutException;
+use Pim\Behat\Decorator\ElementDecorator;
+
+/**
+ * @author    Adrien Pétremann <adrien.petremann@akeneo.com>
+ * @copyright 2016 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class ButtonDecorator extends ElementDecorator
```
Comment (juliensnz) [inline features/Pim/Behat/Decorator/Navbar/ButtonDecorator.php]: this is not the goal of the decorators. The only method for a button decorator could be rightClick for example or move but not asynchronous save. Decorator are made to extend the capabilities of mink, not to implement `asynchronousSave` actions

File: features/Pim/Behat/Decorator/Navbar/ButtonDecorator.php [hunk 22]
```diff
@@ -0,0 +1,67 @@
...
+{
+    use SpinCapableTrait;
+
+    protected $selectors = [
+        'Button' => ['css' => 'button:contains("%s")'],
```
Comment (juliensnz) [inline features/Pim/Behat/Decorator/Navbar/ButtonDecorator.php]: in a decorator we should not search in a page. A button decorator should already be a button element

File: features/Pim/Behat/Decorator/Navbar/ButtonDecorator.php (hunk 21 above)
Comment (juliensnz) [inline features/Pim/Behat/Decorator/Navbar/ButtonDecorator.php]: this class is not a button decorator

File: src/Oro/Bundle/ConfigBundle/Resources/config/form_extensions.yml [hunk 23]
```diff
@@ -3,7 +3,7 @@ extensions:
         module: oro/form/system/config
 
     oro-system-config-tabs:
-        module: pim/product-edit-form/form-tabs
+        module: pim/form/common/form-tabs
```
Comment (juliensnz) [inline src/Oro/Bundle/ConfigBundle/Resources/config/form_extensions.yml]: nice catch

File: features/Context/Page/VariantGroup/Edit.php (hunk 2 above)
Comment (grena) [inline features/Context/Page/VariantGroup/Edit.php]: Yep off course! We didn't clean for now

File: features/Pim/Behat/Decorator/Attribute/AttributeAdderDecorator.php (hunk 11 above)
Comment (grena) [inline features/Pim/Behat/Decorator/Attribute/AttributeAdderDecorator.php]: :+1:

File: features/Pim/Behat/Decorator/Navbar/ButtonDecorator.php (hunk 21 above)
Comment (grena) [inline features/Pim/Behat/Decorator/Navbar/ButtonDecorator.php]: We put it in the Navbar namespace, it's a NavbarButtonsDecorator, it handles navbar button actions no?

File: features/Context/AssertionContext.php (hunk 1 above)
Comment (grena) [inline features/Context/AssertionContext.php]: We'll not use PageDecorator in this story

File: features/Context/Page/VariantGroup/Edit.php (hunk 2 above)
Comment (grena) [inline features/Context/Page/VariantGroup/Edit.php]: We need to keep it for old form fields in order to fallback on the old method

File: features/Context/Page/VariantGroup/Edit.php (hunk 3 above)
Comment (grena) [inline features/Context/Page/VariantGroup/Edit.php]: As seen above, no PageDecorator in this PR

File: features/Pim/Behat/Decorator/Attribute/AttributeAdderDecorator.php (hunk 10 above)
Comment (grena) [inline features/Pim/Behat/Decorator/Attribute/AttributeAdderDecorator.php]: done

File: features/Pim/Behat/Decorator/InputDecorator.php (hunk 15 above)
Comment (grena) [inline features/Pim/Behat/Decorator/InputDecorator.php]: done

File: features/Pim/Behat/Decorator/Navbar/ButtonDecorator.php (hunk 21 above)
Comment (grena) [inline features/Pim/Behat/Decorator/Navbar/ButtonDecorator.php]: we removed the save method, there is only click action now

**Comment by grena** [comment]
Replaced by #4285

**Pull request closed without merging**